reps/2023-07-28-serve-grpc-proxy.md

```diff
@@ -0,0 +1,501 @@
+  and `Router` to be used for gRPC requests. It contains a `grpc_user_request`
+  field to store the byte input for the replica and a `grpc_proxy_handle` field
+  to store the ray serve handle to be used for gRPC requests.
+- `ServeRequest` is a new data model served as the input to `proxy_request()`
+  method so any types of proxy can share the same GA entry point.
+- `ASGIServeRequest` is a subclass of `ServeRequest` used by `HTTPProxy`.
+  It contains the implementations to extract attributes from ASGI input objects for
+  sending requests to the replica
+- `gRPCServeRequest` is a subclass of `ServeRequest` used by `gRPCProxy`.
+  It contains the implementations to extract attributes from gRPC input objects
+  for sending requests to the replica
+- `ServeResponse` is a new data model served as the output of `proxy_request()`
+  method. It has required the string `status_code` used for metrics tracking. It also
+  has optional `response` and `streaming_response` allowing gRPC to pass response
+  back to the client
+
+### gRPCOptions
+This is a new config option we added to the serve config. It contains two fields:
+- `port`: TPort for gRPC server. Will only start gRPC server if set. Cannot be
+  updated once Serve has started running. Serve must be shut down and restarted
+  with the new port instead.
+- `grpc_servicer_functions`: The servicer functions used to add the method handlers
+  to the gRPC server. Default to empty list, which means no gRPC methods will be
+  added and no gRPC server will be started. The servicer functions need to be
+  importable from the context of where Serve is running.
+
+If Serve is started with `gRPCOptions` provided, the `HTTPProxyActor` will start
+the gRPC proxy with the provided port and the provided servicer functions.
+
+#### gRPC Util
+This is where our implementation of gRPC the serialization and deserialization lives.
+We created our custom function `create_serve_grpc_server()` and custom class
+`gRPCServer` to serialize and deserialize user defined protobuf.
+- `create_serve_grpc_server()` is used as a factory function to create custom
+  `gRPCServer` instance. The function will take a method handler factory and
+  the factory will create a callable `unary_unary()` for unary request and
+  `unary_stream()` for streaming request.
+- `gRPCServer` is subclass from `grpc.Server` and it overrides
+  `add_generic_rpc_handlers()` to use our method handlers and to disable
+  `response_serializer` so gRPC proxy can return raw user defined protobuf bytes to
+  the user.
+
+#### Benchmark
+We will benchmark the performance of the gRPC proxy against the existing HTTP proxy
+with new scripts.
+
+#### Docs
+We will add new docs and code examples on how to use gRPC proxy.
+
+#### Dependencies
+We previously implemented the direct ingress with `gRPCIngress` using Ray's default
+gRPC library. We will need to include gRPC library to Serve's default libraries
```

Comment:
gRPC is quite heavy-weight library bringing a lot of deps, bringing additional potential for conflicts. Can we make it part of standalone module `ray[serve-grpc]` module?

Reply:
Hmm, currently there is already a `gRPCIngress` implementation and the new `gRPCProxy` doesn't take anymore dependencies than that (no additional dependency required). But for sure we can think about adding a barebone serve installation without gRPC if we see fit.